AP_Motors:Heli_RSC Add heli rotor speed governor

ChristopherOlson opened the pull request:
This adds a rotor speed governor to the Heli_RSC. It completes the open/closed loop throttle control modes for the Heli_RSC. It has been extensively flight tested over several months. It is primarily intended for piston engine helicopters, or Wren 44i/50i turbine engine helicopters that do not have a governor in their FADEC. But has also been tested on electric.

It also adds logging for the governor and throttle output.

lvale commented:
@ChristopherOlson we need a wiki entry for setting up. Also would be great if we should document this on Electrics that the ESC does not have a Governor function. Also on the wiki, if a ESC has governor, which would be the pros and cons of having this, or the ESC governor.

ChristopherOlson commented:
@lvale indeed wiki documentation still needs to be done. As well as for other features like the new swashplate setup that is coming with four-servo support. The governor is very easy to set up, and therefore quite easy to document for the wiki, a number of users are already flying it on 3.5 and 3.6. It is merely an added mode in the Heli_RSC, and extension of the throttle curve.

The swashplate stuff will be more extensive to document.

ChristopherOlson commented on libraries/AP_Motors/AP_MotorsHeli_Dual.cpp:
```diff
@@ -265,6 +271,10 @@ void AP_MotorsHeli_Dual::calculate_armed_scalars()
     _rotor.set_throttle_curve(thrcrv, (uint16_t)_rsc_slewrate.get());
+    _rotor.set_governor_disengage(_rsc_governor_disengage*0.01f);
+    _rotor.set_governor_droop_setting(_rsc_governor_droop_setting*0.01f);
+    _rotor.set_governor_setpoint(_rsc_governor_setpoint);
```
The setpoint for the RSC is a PWM value for ESC's. The governor setpoint uses actual rotor rpm. Since they are two essentially different things I made them separate.

ChristopherOlson replied:
All these estimated speeds, including critical speed, are not governor values - they are pwm values and governors deal with rpm. That is one of the reasons why I added the _governor_status variable that switches from 0 to 1 when the governor is at full operational output, meaning the difference between the reference and setpoint is within the operational droop limit of the governor.

None of the principles of operation of a droop speed control apply to, nor can use, any of the RSC stuff except for the throttle curve. Because the rest of it all based on pwm values for throttle.

Droop speed controls have three components:
1) Setpoint
2) Reference Speed
3) Fuel control
In ArduPilot the Setpoint is the desired governor speed. The reference speed is the measured rpm value. The fuel control is throttle output. Droop is the difference between setpoint and reference speed. Reference speed goes down with increasing load, and this is what provides the stability of a droop speed control. Fuel control is proportional to droop, either linear, sqrt, or x^2 depending on the governor design.

The ramp time is still used to ramp the throttle, the runup time is still used to hold things until the _governor_status goes to 1. But just like any of the other modes, the ramp time must be determined by the end user because not all engines and rotor systems accelerate at the same rate. That was the primary reason a timer does not work to soft-start a droop governor. A percentage of the fuel control works the best for soft-starting, since fuel control is directly related to the amount of torque applied vs throttle opening. Maximum torque is achieved at WOT and peak BMEP, etc.. And the power (horsepower) relationship is directly linked to torque and rpm over time (hp = (torque * rpm) / 5252. So using this principle of how much fuel control is applied predicts the acceleration speed quite accurately for all engine types, based on a properly tuned throttle curve, that fits the prime mover's torque curve.

So in a nutshell that is the basics of how a droop speed control works. How to integrate that with the existing stuff in the RSC for other modes, I'm at a loss. Most of the rest of it was designed for electrics. This is the first ever fuel engine control mode in Copter.

bnsgeyer replied:
I’m just saying it doesn’t matter what the variable represents PWM or rpm. As long as the variable is of the same type then they can be interchangeable. And it is not like your using two RSC modes at the same time.

ChristopherOlson replied:
So you're saying we could use current H_RSC_SETPOINT set to PWM value for electric ESC, but change it to rpm value for Mode 4 governor? In that case, what do we use for a default?

ChristopherOlson replied:
Ok, that will work. I'm working on rebasing this and re-ordering the commits so it can be more easily pushed to for further changes. As soon as I get that done I'll do a force push to my branch and it should be cleaned up. I left the commits to Copter individual, but look it over after I push it, I could squash those Copter commits too. Unless there is an issue with the call in Heli.cpp for the rpm, which will now be squashed into the AP_Motors. We may want to move that call to the RSC?

ChristopherOlson commented:
@bnsgeyer Bill, look this over now - I think I have it in a suitable order, and cleaned up so you can make the changes you want to make to the params, etc.. I hope I did this right. Apologies for doing a hasty port to 3.7 from the working code.

bnsgeyer commented:
> @bnsgeyer Bill, look this over now - I think I have it in a suitable order, and cleaned up so you can make the changes you want to make to the params, etc.. I hope I did this right. Apologies for doing a hasty port to 3.7 from the working code.

@ChristopherOlson  I looked at your revised branch and you still had a change to the heli.cpp in your commit for the AP_Motors library.  So I went and created a branch in my repo and cherry-picked the original commits from your 3.6 branch.  I was able to to do it so that I didn't lose your authorship on the commits.  I omitted the change for setting rpm from heli.cpp.  So I spent a lot of time last night trying to figure out how to get the rpm sensor data without setting it from the greater code.  I think I have it figured out.  I have a question into Randy and hopefully he'll confirm my proposed solution.  Once I figure that out, I will start making the other changes.  We may want to talk at some point about this critical speed parameter and its role in each of the rsc modes.  I'm currently using it to switch from spool down to ground idle.  when it knows the actual rotor speed, the rsc could use actual rotor speed to signal transition. just a thought.

ChristopherOlson commented:
Yes, I did not make any changes to the original code. Just cleaned up the commit order and squashed a couple to make it cleaner so you can make the changes you want.

I could not figure out how to get rpm into the RSC, so that's why I did it in heli.cpp. I could not get the governor to work until I did that. Since it was just one line of code I included it in the governor commit instead of separating it. At present, the governor doesn't work without that line of code.

ChristopherOlson commented:
@bnsgeyer I made the changes we discussed, including using a boolean flag for governor status, and fixing up the set_rpm call for all frame types. Then rebased on the lastest master.

IMO, further changes to how the setpoint vs governor rpm is used in the code should be addressed in a further overhaul of the RSC. These things are two essentially different methods at present. Adding in the governor for closed-loop control identifies these issues that should be fixed in a RSC re-write. But those changes would be not be directly related to adding in the governor, as it affects other things than just the governor function.

ChristopherOlson commented on ArduCopter/heli.cpp:
```diff
@@ -159,6 +159,7 @@ void Copter::heli_update_rotor_speed_targets()
             // pass setpoint through as desired rotor speed, this is almost pointless as the Setpoint serves no function in this mode
             // other than being used to create a crude estimate of rotor speed
             if (motors->get_interlock()) {
+                ((AP_MotorsHeli_Single&)*motors).set_rpm(rpm_sensor.get_rpm(0.0f));
```
This has been changed.

ChristopherOlson commented on ArduCopter/Log.cpp:
```diff
@@ -440,7 +440,7 @@ const struct LogStructure Copter::log_structure[] = {
 #if FRAME_CONFIG == HELI_FRAME
     { LOG_HELI_MSG, sizeof(log_Heli),
-      "HELI",  "Qff",         "TimeUS,DRRPM,ERRPM", "s--", "F--" },
+      "HELI",  "Qff",        "TimeUS,DRRPM,ERRPM", "s--", "F--" },
```
this has also been fixed

bnsgeyer commented:
@ChristopherOlson yes what you've done looks good.  I still want to break out the parameters and only have them enabled for the particular RSC mode.  We are getting too many parameters stacking up in heli and it won't take long to incorporate this into your PR.  Those running RSC setpoint don't need to see throttle curve and Governor settings.  I would have done this back when we put the throttle curve in but I'm not sure it was available.  I will work it tonight and have a commit you can pull in.  Sorry I've stalled progress on this.

ChristopherOlson commented:
OK, great. I wasn't sure if we were going to do that as part of the governor, or in a more sweeping overhaul of the RSC. The ground stations will have to programmed for this, as I noticed with the swashplate testing that the disappearing params cause Mission Planner to "hang" looking for params that aren't there.

ChristopherOlson commented:
@bnsgeyer I reviewed your changes, pulled them into my branch, built it, tested it in a real helicopter and everything seemed to work. So I rebased on master again and pushed it to update the PR.

ChristopherOlson commented:
I rebased this again on master, and squashed some of the commits.

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -69,7 +69,7 @@ class AP_MotorsHeli_RSC {
     
     // set_rotor_rpm - when speed sensor is available for governor
-    void        set_rotor_rpm(int16_t rotor_rpm) {_rotor_rpm = rotor_rpm; }
+    void        set_rotor_rpm(int16_t rotor_rpm) {_rotor_rpm = (float)rotor_rpm; }
```
we could probably make the rotor_rpm argument a float.

tridge commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -113,3 +152,47 @@ class AP_MotorsHeli_RSC {
+    static const struct AP_Param::GroupInfo var_info[];
+
+    void set_thrcrv_enable(int8_t setenable) {enable = setenable; }
+    float * get_thrcrv() {
```
this is a bit odd ... seems to be doing a cast via a static array

ChristopherOlson replied:
Now I see what you were referring to. It was confusing at first, as this part of the code had not been changed in this PR.

rmackay9 commented:
As mentioned I'll definitely do a detailed review before next week's call.

rmackay9 commented on ArduCopter/ArduCopter.cpp:
```diff
@@ -152,7 +152,7 @@ const AP_Scheduler::Task Copter::scheduler_tasks[] = {
     SCHED_TASK_CLASS(AP_Scheduler,         &copter.scheduler,           update_logging, 0.1,  75),
 #if RPM_ENABLED == ENABLED
-    SCHED_TASK(rpm_update,            10,    200),
+    SCHED_TASK(rpm_update,            40,    200),
```
This will lead to the RPM data being logged at 40hz as well.  Perhaps that's OK but I just want to be sure you're aware.

ChristopherOlson replied:
Yes, we are aware of this. Discussed it with tridge in the early stages of governor development. He said it is a very lost cost change from a resources standpoint. The five-element mode filter in the rpm library sort of slows things down a bit and we're reducing the latency as much as possible to get real-time rpm reading here.

rmackay9 commented on ArduCopter/Log.cpp:
```diff
@@ -403,7 +407,7 @@ const struct LogStructure Copter::log_structure[] = {
 #if FRAME_CONFIG == HELI_FRAME
     { LOG_HELI_MSG, sizeof(log_Heli),
-      "HELI",  "Qff",         "TimeUS,DRRPM,ERRPM", "s--", "F--" },
+      "HELI",  "Qffff",        "TimeUS,DRRPM,ERRPM,GOV,THROT", "s----", "F----" },
```
No big deal of course but most logging column names are a mix of lower and upper case

ChristopherOlson replied:
I will fix this, as well.

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli.cpp:
```diff
@@ -77,7 +77,7 @@ const AP_Param::GroupInfo AP_MotorsHeli::var_info[] = {
     // @DisplayName: Rotor Speed Control Mode
     // @Description: Determines the method of rotor speed control
-    // @Values: 1:Ch8 Input, 2:SetPoint, 3:Throttle Curve
+    // @Values: 1:Ch8 Input, 2:SetPoint, 3:Throttle Curve, 4: Governor
```
extra little space between "4:" and "Governor"

ChristopherOlson replied:
Oops! I'll fix that.

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli.cpp:
```diff
@@ -193,6 +193,38 @@ const AP_Param::GroupInfo AP_MotorsHeli::var_info[] = {
     // @User: Standard
     AP_GROUPINFO("RSC_THRCRV_100", 24, AP_MotorsHeli, _rsc_thrcrv[4], AP_MOTORS_HELI_RSC_THRCRV_100_DEFAULT),
 
+    // @Param: RSC_GOV_SET
```
It might be a good idea to move the RSC_ parameters off into a separate group so they appear/disappear if the feature is enabled/disabled.  This is an existing thing though.

rmackay9 replied:
this is done in a follow-up commit so marking as "resolved".

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -178,16 +181,18 @@ class RSCGovFloatParam {
-    AP_Float  tc;           // governor throttle curve weighting, range 50-100%
+    AP_Int16  reference;      // sets rotor speed for governor
+    AP_Float  range;          // RPM range +/- governor rpm reference setting where governor is operational
+    AP_Float  disengage;      // sets the throttle percent where the governor disengages for return to flight idle
```
Perhaps "disengage" should be renamed to something more specific.  like thr_disengage_pct or something like that.

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -178,16 +181,18 @@ class RSCGovFloatParam {
+    AP_Float  range;          // RPM range +/- governor rpm reference setting where governor is operational
+    AP_Float  disengage;      // sets the throttle percent where the governor disengages for return to flight idle
+    AP_Float  droop_response; // governor response to droop under load
+    AP_Float  tc;             // governor throttle curve weighting, range 50-100%
```
perhaps "tc" could also be made more specific.  I would normally assume "TC" meant "time constant".  maybe "thr_curve_weighting".  Doesn't need to be really long but a bit longer to help the next developer guess what it's for without looking at the declaration comments would be good.

ChristopherOlson replied:
This is now declared as _governor_thrcurve

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -178,16 +181,18 @@ class RSCGovFloatParam {
-    AP_Float  droop_setting;// governor response to droop under load
-    AP_Float  tc;           // governor throttle curve weighting, range 50-100%
+    AP_Int16  reference;      // sets rotor speed for governor
+    AP_Float  range;          // RPM range +/- governor rpm reference setting where governor is operational
```
maybe "range" should be "rpm_range" or "range_rpm"?

ChristopherOlson replied:
This would be an easy enough change. But do we really want to make the variable names longer when the variable is declared as '_governor_range' with a clear explanation in the code comment?

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -178,16 +181,18 @@ class RSCGovFloatParam {
-    AP_Int16  setpoint;     // sets rotor speed for governor
-    AP_Float  droop_setting;// governor response to droop under load
-    AP_Float  tc;           // governor throttle curve weighting, range 50-100%
+    AP_Int16  reference;      // sets rotor speed for governor
```
"reference" could be "target_rpm"?

ChristopherOlson replied:
On this one, the term "reference" was taken from Woodward's engineering whitepapers on droop governors. It is called a 'reference' because it is not a setpoint. The governor moves the reference, as needed, to compensate for load so it is constantly a moving target when the governor is operating.

See:http://www.woodward.com/WorkArea/DownloadAsset.aspx?id=2147483661

ChristopherOlson replied:
On this one I'm not comfortable with changing the variable name due to the design of this type of governor. It truly is a _governor_reference and not a 'target_rpm'. If the governor reference is 1500 and it droops by 20 it will move the reference to 1520 by adding droop. A droop governor never actually achieves such a thing as a 'target_rpm'. It always operates in droop, with droop compensation by moving the reference.

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_Single.cpp:
```diff
@@ -231,8 +231,9 @@ void AP_MotorsHeli_Single::calculate_armed_scalars()
-        _main_rotor.set_governor_droop_setting(_rsc_gov.get_droop_setting()*0.01f);
-        _main_rotor.set_governor_setpoint(_rsc_gov.get_setpoint());
+        _main_rotor.set_governor_droop_response(_rsc_gov.get_droop_response()*0.01f);
+        _main_rotor.set_governor_reference(_rsc_gov.get_reference());
```
"reference" is a new word.. which is OK but normally we use "target".  I'm actually not a fan of the word "setpoint" but "reference" isn't too clear either I think.  It's up to you guys but I think keeping terminology consistent helps developers move more easily between different parts of the code.

ChristopherOlson replied:
This seems to be a duplicate comment on the use of the 'reference' in a droop governor. I would prefer to keep it as designed, as it is not the same as a "target".

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -178,16 +181,18 @@ class RSCGovFloatParam {
+    float get_range() { return range; }
     float get_disengage() { return disengage; }
-    float get_droop_setting() { return droop_setting; }
+    float get_droop_response() { return droop_response; }
```
I wonder if this "response" is really a "gain"?

ChristopherOlson replied:
Not really. It is the governor's droop response setting as a percentage of the max allowable droop. It could be considered a type of gain, but since this is a droop speed control, and not a PID control, the droop setting determines the governor's response to the fuel control. It is the droop setting that moves the governor's reference speed and determines the governor's stability without going out on over-speed.

see (on page 5):http://www.woodward.com/WorkArea/DownloadAsset.aspx?id=2147483661

I elected to use this type of speed control over a PID due to the droop governor's fantastic stability without surging. Since it is a different type of design that commonly seen in ArduPilot some of the terminology used will be a bit confusing or different than a typical PID control.

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -131,9 +133,10 @@ class AP_MotorsHeli_RSC {
-    float           _governor_droop_setting = 0.0f; // governor droop setting, range 0-100%
     float           _governor_output = 0.0f;        // governor output for rotor speed control
-    float           _governor_setpoint = 0.0f;      // governor speed setpoint, range 800-3500 rpm
+    float           _governor_range = 0.0f;         // RPM range +/- governor rpm reference setting where governor is operational
```
There's actually no need to initialise these variables to zero I hear.  It just makes the code size a little bit larger than it needs to be.

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli.cpp:
```diff
@@ -153,77 +153,15 @@ const AP_Param::GroupInfo AP_MotorsHeli::var_info[] = {
-    AP_GROUPINFO("RSC_GOV_DROOP", 27, AP_MotorsHeli, _rsc_governor_droop_setting, AP_MOTORS_HELI_RSC_GOVERNOR_DROOP_DEFAULT),
+    // @Group: RSC_CRV_
+    // @Path: AP_MotorsHeli_RSC.cpp
+    AP_SUBGROUPINFO(_rsc_thrcrv, "RSC_CRV_", 27, AP_MotorsHeli, RSCThrCrvInt16Param),
```
this is very nice to see.  I'll delete my previous comment which was a suggestion to do this!

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -20,6 +20,112 @@
+    AP_GROUPEND
+};
+
+const AP_Param::GroupInfo RSCGovFloatParam::var_info[] = {
```
"RSCGovFloatParam" is OK but we could probably lose the "Float" part.  The class doesn't only include Floats and I don't think we have a Float version and an Int version or anything so the distinction isn't really ncessary perhaps?

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.cpp:
```diff
@@ -20,6 +20,112 @@
 
 extern const AP_HAL::HAL& hal;
 
+const AP_Param::GroupInfo RSCThrCrvInt16Param::var_info[] = {
```
Similarly here I think we could lose the "Int16" part of the class name.

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -5,6 +5,22 @@
+#define AP_MOTORS_HELI_RSC_SETPOINT             700
+
+// Throttle Curve Defaults
+#define AP_MOTORS_HELI_RSC_THRCRV_0_DEFAULT     250
```
Peter Barker and others would say that we should be moving to use enum classes instead of defines because it adds extra type safety.  It's not a blocker but in the future moving to enums is a good idea.

rmackay9 commented on libraries/AP_Motors/AP_MotorsHeli_RSC.h:
```diff
@@ -133,3 +152,45 @@ class AP_MotorsHeli_RSC {
+
+private:
+    AP_Int8   enable;
+    AP_Float  disengage;    // sets the throttle percent where the governor disengages for return to flight idle
```
I mentioned earlier as well but perhaps we could make the parameter variables names a bit more specific to help future developers?

ChristopherOlson commented:
@rmackay9 I think I have most of your suggestions addressed? There is still a few from commits that @bnsgeyer made to this project that I did not address. Could you take a look those Bill?

Rather than making a bunch of separate new commits I squashed them where possible to make the entire PR a little cleaner.

bnsgeyer commented:
@ChristopherOlson I made the changes for the names of the parameter groups.  It is in a copy of your branch on my repo.  [here](https://github.com/bnsgeyer/ardupilot/tree/olson-pr-heli-governor)
You can cherry pick it into your branch.  I don't like pushing to another users branch.  Is that all you wanted me to look at/fix.  Let me know.

ChristopherOlson commented:
@bnsgeyer I will pull these in and re-push to update the PR. I THINK this is what Randy wanted but I have to review the rest of his comments to see. Some of the things, like making the variable names longer, I don't agree with. IMO the variables are well commented and due to the method used to hide params they may seem a bit cryptic. But where the variable is declared it should be perfectly obvious as being a governor or _gov variable.

ChristopherOlson commented:
@rmackay9 I think we have your comments addressed mostly. I pushed the final commit to change the rotor_rpm argument to a float value. And I think @bnsgeyer changed what was requested in the commits he made. If you get a chance, please re-review?

bnsgeyer commented:
@christopherolson I only pushed to my branch. I wanted you to cherry pick it in. Here’s the commit 
https://github.com/bnsgeyer/ardupilot/commit/8af3e6bc87fc29324a55e543cf665868b6eebd12

ChristopherOlson commented:
@bnsgeyer yes, I pulled that in and it is in the commits on this PR. I think that was the extent of what Randy requested for changes on the commits you had made.

bnsgeyer commented:
Sorry. Didn’t look far enough back in the thread. It will be good to get this in.

rmackay9 commented:
Merged, thanks!

MidwestAire commented:
I originally did that to make the helicopter rotor governor work properly. I initially tried it at 10Hz and could not get a fast enough update on rpm for the governor to work properly. So I increased it to 40Hz and got the desired result. The change was deemed to be not excessively resource intensive so was left that way for Copter.
